test/bntests.pl

```diff
@@ -0,0 +1,162 @@
+        return if $lshift1 == $a->bmul(2);
+    } elsif ( defined $s{'LShift'} ) {
+        # LShift = A * 2**N
+        my $lshift = &bn($s{'LShift'});
+        my $a = &bn($s{'A'});
+        my $n = &bn($s{'N'});
+        return if $lshift == $a->blsft($n);
+    } elsif ( defined $s{'RShift'} ) {
+        # RShift = A / 2**N
+        my $rshift = &bn($s{'RShift'});
+        my $a = &bn($s{'A'});
+        my $n = &bn($s{'N'});
+        return if $rshift == $a->brsft($n);
+    } elsif ( defined $s{'Square'} ) {
+        # Square = A * A
+        my $square = &bn($s{'Square'});
+        my $a = &bn($s{'A'});
+        return if $square == $a->bmul($a);
+    } elsif ( defined $s{'Product'} ) {
+        # Product = A * B
+        my $product = &bn($s{'Product'});
+        my $a = &bn($s{'A'});
+        my $b = &bn($s{'B'});
+        return if $product == $a->bmul($b);
+    } elsif ( defined $s{'Quotient'} ) {
+        # Quotient = A / B
+        # Remainder = A - B * Quotient
+        my $quotient = &bn($s{'Quotient'});
+        my $remainder = &bn($s{'Remainder'});
+        my $a = &bn($s{'A'});
+        my $b = &bn($s{'B'});
+
+        # First the remainder test.
+        $b->bmul($quotient);
+        my $rempassed = $remainder == $a->bsub($b) ? 1 : 0;
+
+        # Math::BigInt->bdiv() is documented to do floored division,
+        # i.e. 1 / -4 = -1, while OpenSSL BN_div does truncated
+        # division, i.e. 1 / -4 = 0.  We need to make the operation
+        # work like OpenSSL's BN_div to be able to verify.
+        $a = &bn($s{'A'});
+        $b = &bn($s{'B'});
+        my $neg = $a->is_neg() ? !$b->is_neg() : $b->is_neg();
+        $a->babs();
+        $b->babs();
+        $a->bdiv($b);
+        $a->bneg() if $neg;
+        return if $rempassed && $quotient == $a;
+    } elsif ( defined $s{'ModMul'} ) {
+        # ModMul = (A * B) mod M
+        my $modmul = &bn($s{'ModMul'});
+        my $a = &bn($s{'A'});
+        my $b = &bn($s{'B'});
+        my $m = &bn($s{'M'});
+        $a->bmul($b);
+        return if $modmul == $a->bmod($m);
+    } elsif ( defined $s{'ModExp'} ) {
+        # ModExp = (A ** E) mod M
+        my $modexp = &bn($s{'ModExp'});
+        my $a = &bn($s{'A'});
+        my $e = &bn($s{'E'});
+        my $m = &bn($s{'M'});
+        return if $modexp == $a->bmodpow($e, $m);
+    } elsif ( defined $s{'Exp'} ) {
+        my $exp = &bn($s{'Exp'});
+        my $a = &bn($s{'A'});
+        my $e = &bn($s{'E'});
+        return if $exp == $a ** $e;
+    } elsif ( defined $s{'ModSqrt'} ) {
+        # ModSqrt * ModSqrt = A mod P
+        my $modsqrt = &bn($s{'ModSqrt'});
+        my $a = &bn($s{'A'});
+        my $p = &bn($s{'P'});
+
+        # With OpenSSL's BN, the result of -10 % 3 is -1
+        # while Math::BigInt, the result is 2.
+        # Math::BigInt does something different with a negative modulus,
+        # while OpenSSL's BN and bc treat it like a positive number...
+        $a->babs();
+        $p->babs();
```

Comment:
If you remove the `babs` calls, the expected failures are down at 50.  Like I discussed already, `BN_mod_sqrt()` uses `BN_nnmod()`, which *does* behave like `Math::BigInt` modulo, so no extra hacking needed here.